server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,292 @@
+            self._logger.warning("given searches are impossible to split in even teams because of party sizes")
+            raise UnevenTeamsException
+        candidates = SortedList(searches_dict[1], key=lambda s: s.cumulated_rating)
+
+        if avg - team_avg < 0:
+            iterator = iter(candidates)
+        else:
+            iterator = iter(candidates[::-1])
+        candidate = next(iterator)
+        for item in iterator:
+            team_avg = get_average_rating([search, candidate])
+            avg_delta = abs(avg - team_avg)
+            new_team_avg = get_average_rating([search, item])
+            new_avg_delta = abs(avg - new_team_avg)
+            self._logger.debug("delta with %s is %s (avg is %s)", [candidate], avg_delta, team_avg)
+            self._logger.debug("new delta with %s is %s (avg is %s)", [item], new_avg_delta, new_team_avg)
+            if new_avg_delta > avg_delta:
+                break
+            candidate = item
+        self._logger.debug("used %s as filler", [candidate])
+        return candidate
+
+    def calculate_game_quality(self, match: Match) -> Game:
+        newbie_bonus = 0
+        time_bonus = 0
+        ratings = []
+        for team in match:
+            time_bonus += team.failed_matching_attempts * config.TIME_BONUS
+            time_bonus = min(time_bonus, config.MAXIMUM_TIME_BONUS)
+            newbie_bonus += team.has_newbie() * config.NEWBIE_BONUS
+            for search in team.get_original_searches():
+                ratings.append(search.average_rating)
+
+        rating_disparity = abs(match[0].cumulated_rating - match[1].cumulated_rating)
+        fairness = max((config.MAXIMUM_RATING_IMBALANCE - rating_disparity) / config.MAXIMUM_RATING_IMBALANCE, 0)
+        deviation = statistics.pstdev(ratings)
+        uniformity = max((config.MAXIMUM_RATING_DEVIATION - deviation) / config.MAXIMUM_RATING_DEVIATION, 0)
+
+        quality = fairness * uniformity + newbie_bonus + time_bonus
+        self._logger.debug(
+            "bonuses: %s rating disparity: %s -> fairness: %f deviation: %f -> uniformity: %f -> game quality: %f",
+            newbie_bonus + time_bonus, rating_disparity, fairness, deviation, uniformity, quality)
+        return Game(match, quality)
+
+    def _pick_best_noncolliding_games(self, games: List[Game]) -> List[Match]:
+        for game in list(games):
+            if game.quality < config.MINIMUM_GAME_QUALITY:
+                games.remove(game)
+        self._logger.debug("%i games left after removal of games with quality < %s",
+                           len(games), config.MINIMUM_GAME_QUALITY)
+        games = SortedList(games, key=lambda gme: gme.quality)
+
+        matches: List[Match] = []
+        while len(games) > 0:
+            g = games.pop()
+            matches.append(g.match)
+            used_players = set()
+            for search in g.match:
+                for player in search.players:
+                    used_players.add(player)
```

Comment:
This can become a comprehension (technically it's a function call with a generator expression). I think something like this is the correct order for nested loops:
```suggestion
            used_players = set(player for search in g.match for player in search.players)
```